refactor: Refactored ui.tabs component design (#504)
deephaven/deephaven-plugins: plugins/ui/DESIGN.md

)
```

###### ui.tabs

A tabs component can be used to organize content in a collection of tabs, allowing users to navigating between the different tabs. Children (the tabs) can be specified in one of two ways:

1. They can be of type `Item`, where each item represents an individual tab panel

> 2. They can be of types `TabList` and `TabPanels`, which when combined, outline all tabs and the respective contents

Reviewer: Why would you use one over the other? Mention in the second case that you may want to use that case to have more control over the layout of the panels/lists.

| Parameter               | Type                                  | Description                                                                                                                                                    |
| ----------------------- | ------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `*children`             | `Item \| TabList \| TabPanels`        | The tab panels to render within the tabs component.                                                                                                                                                                                               |
| `on_change`             | `Callable[[Key], None] \| None` | Alias of `on_selection_change`. Handler that is called when the tab selection changes.                                                                            |
| `**props`               | `Any`                                 | Any other [Tabs](https://react-spectrum.adobe.com/react-spectrum/Tabs.html) prop 
|

```py
from deephaven import empty_table, ui

> from deephaven.plot import express as dx

Reviewer: ```suggestion
```

> # Elementary Cases

Reviewer: Instead of one large code-block, these should be separate code blocks with text description in-between them. E.g.

###### Tabs using `ui.item`

If you're just using the default tab layout and don't need to customize the appearance of the tabs, you can simply pass in `ui.item` directly to `ui.tabs` as a concise method of defining your tabs.
```python
ui.tabs(
  ui.item(...),
  ui.item(...),
)
```
... description of another ui.item case, and another code block, etc.

###### Tabs using `ui.tab_panels` and `ui.tab_list`

If you need to customize the appearance/layout of the tab headers/content, then use `ui.tab_list` and `ui.tab_panels` to allow for extensive customization.

```python
ui.tabs(
  ui.tab_list(...),
  ui.tab_panels(...)
)
```

###### Error cases

ui.tabs(
    # Title and content passed in as string literals

>     # Should render a tab with the title "Tab 1" with "Content 1" as tab content, that given no key passed, would be set to "Tab 1 Key"

Reviewer: Docs should be written in the imperative mood; no need to say "Should render", just say "Render".
```suggestion
    # Render a tab with the title "Tab 1" with "Content 1" as tab content, that given no key passed, would be fallback to the title "Tab 1" as the key
```

    ui.item("Content 1", title="Tab 1"),

    # Title, key, and content passed in as string literals
    # Should render a tab with the title "Tab 2" with "Content 2" as tab content, keyed "Key 2"
    ui.item("Content 2", title="Tab 2", key="Key 2"),

    # Content passed in that contains a flex, illustrating that tab content does not have to be a string

>     # Should render a tab with the title "Tab 3" with "Hello World" header with a table beside it as the tab content

Reviewer: ```suggestion
    # Render a tab with the title "Tab 3" with "Hello World" header with a table beside it as the tab content
```

    # Given that the key is not provided, the title "Tab 3" defaults as the key
    ui.item(
        ui.flex(
            "Hello World!",
            ui.flex(empty_table(10).update("I=i")),
        ),
        title="Tab 3",
        key="Key 3",
    ),

    # Tab with text and icon as title

>     # Should render "Content 4" in a tab called "<GITHUB LOGO> Tab 4", keyed "<GITHUB LOGO> Tab 4 Key"
>     ui.item("Content 4", title=ui.item("Tab 3", ui.icon("vsGithubAlt")))

Reviewer: Your description does not match the code. Also the key cannot contain an icon; the key will just fallback to the text value.
```suggestion
    # Render "Content 4" in a tab called "<GITHUB LOGO> Tab 4". With no key provided, it will fallback to the text content of the title, "Tab 4".
    ui.item("Content 4", title=ui.item(ui.icon("vsGithubAlt"), "Tab 4"))
```

)

# Tab with on_change callback
ui.tabs(

>     # Should render a tabs that have an on_change, which prints the selected tab key when a tab is selected

Reviewer: ```suggestion
    # Render tabs that have an `on_change`, which prints the selected tab key when a tab is selected
```

    ui.item("Content 1", title="Tab 1", key="Key 1"),
    ui.item("Content 2", title="Tab 2", key="Key 2"),
    on_change=lambda key: print(f"Selected key: {key}"),
)

# Tabs specified by passing in tab_panels and tab_list
ui.tabs(
    # Should render a Tab with a title of "Tab 1", with a content of "Content 1", keyed "Key 1"
    # But will not, given the raising of a KeyError, since there are mismatching keys for "Tab 2"
    ui.tab_list(ui.item("Tab 1", key="Key 1"), ui.item("Tab 2", key="Key 2")),
    ui.tab_panels(
        ui.item("Content 3", key="Key 1"),
        ui.item("Content 2", key="Key 3"),
        flex_grow=1,
        position="relative",
    ),
    flex_grow=1,

> )

Reviewer: Show a successful case of using `ui.tabs` with `ui.tab_list` and `ui.tab_panels` before showing an erroneous case.

Also make it clear that in the `tab_list`/`tab_panels` case, you must specify the keys and they must match, or you may have an error (as you have described here).

# Some error-causing cases
@ui.component

> def tabs_error_cases():

Reviewer: Not sure why you changed the format up here. As mentioned above, each case should have it's own block; no need to make a `ui.component` with three return's in it (which doesn't make sense)
